bigint/src/lib.rs

```diff
@@ -1349,6 +1356,47 @@ impl Integer for BigUint {
     }
 }
 
+fn high_bits_to_u64(v: &BigUint) -> u64 {
+    match v.data.len() {
+        0   => 0,
+        1   => v.data[0] as u64,
+        _   => {
+            let mut bits = v.bits();
+            let mut ret = 0u64;
+            let mut ret_bits = 0;
+
+            for d in v.data.iter().rev() {
+                let digit_bits = (bits + big_digit::BITS - 1) %
+                    big_digit::BITS + 1;
```

Comment:
Strictly speaking, it's not correct.  It just trades a should-be-impossible subtraction error for a possible-but-unlikely addition overflow.  And neither way gives a reasonable result, because `bits==0` is a corner case where `digit_bits` should be 0, otherwise `bits -= bits_want` will break.  It's not even more readable, because you had to wrap the line to fit the extra addition.